### ISSUE TITLE
Could you share your experience in solving the problem

### ISSUE THREAD

**Issue by KeyKy** [zhreshold/mxnet-ssd#48]
I have seen the issue you post in https://github.com/weiliu89/caffe/issues/202. Could you share your experience in solving the problem?

**Comment by zhreshold** [comment]
I could confirm that the Normalize layer is exactly l2norm+scaling. I also tested adding normalize layer to relu7, with similar results. Since the original model don't have batchNorm layer, so you have to tune the parameters very carefully in order to make it converge.
Hope this could help a bit.

**Comment by KeyKy** [comment]
@zhreshold Do you have FDDB face dataset example? I wrote a fddb.py but the Train-ObjectAcc is too bad(0.058) even after 680 epochs. Here is my [fddb.py]https://github.com/KeyKy/issue/blob/master/fddb.py)

**Comment by KeyKy** [comment]
@zhreshold I have tested your code with my fddb.py and it works! Could you give me some tips to merge your code with the latest mxnet. Thanks~!

**Comment by zhreshold** [comment]
You could use example/ssd in the latest mxnet, I've submitted PR

**Comment by KeyKy** [comment]
@zhreshold I have tested example/ssd + my fddb.py but it can not converge, while this repository + my fddb.py converges.

**Comment by coolbrain** [comment]
@zhreshold  I implement this paper in Mxnet( Inside-Outside Net: Detecting Objects in Context with Skip Pooling and Recurrent Neural Networks[https://arxiv.org/abs/1512.04143](url)), but L2Normalize + scale can't work as well as in Caffe.
The configuration in caffe prototxt:
`layer {
  name: "roi_pool3_norm"
  type: "Normalize"
  bottom: "roi_pool3"
  top: "roi_pool3_norm"
  norm_param {
   across_spatial: false
   scale_filler {
	type: "constant" 
	value: 2
   }
   channel_shared: false
  }
}`

The redefining network code in Mxnet:
`  roi_pool3 = mx.symbol.ROIPooling(
         name='roi_pool3', data=relu3_3, rois=rois, pooled_size=(7, 7), spatial_scale=0.25)
   roi_pool4 = mx.symbol.ROIPooling(
         name='roi_pool4', data=relu4_3, rois=rois, pooled_size=(7, 7), spatial_scale=0.125)
   roi_pool5 = mx.symbol.ROIPooling(
         name='roi_pool5', data=relu5_3, rois=rois, pooled_size=(7, 7), spatial_scale=0.0625)
   roi_pool3_norm = mx.symbol.L2Normalization(data=roi_pool3, name='roi_pool3_norm', mode='channel')
   roi_pool4_norm = mx.symbol.L2Normalization(data=roi_pool4, name='roi_pool4_norm', mode='channel')
   roi_pool5_norm = mx.symbol.L2Normalization(data=roi_pool5, name='roi_pool5_norm', mode='channel')

   roi_pool3_scale = mx.symbol.Variable(name="roi_pool3_scale",
                                          shape=(1, 256, 1, 1), init=mx.init.Constant(2))
   roi_pool4_scale = mx.symbol.Variable(name="roi_pool4_scale",
                                          shape=(1, 512, 1, 1), init=mx.init.Constant(2))
   roi_pool5_scale = mx.symbol.Variable(name="roi_pool5_scale",
                                          shape=(1, 512, 1, 1), init=mx.init.Constant(2))
  roi_pool3_norm_scale = mx.symbol.broadcast_mul(lhs=roi_pool3_scale, rhs=roi_pool3_norm)
  roi_pool4_norm_scale = mx.symbol.broadcast_mul(lhs=roi_pool4_scale, rhs=roi_pool4_norm)
  roi_pool5_norm_scale = mx.symbol.broadcast_mul(lhs=roi_pool5_scale, rhs=roi_pool5_norm)
  roi_pool_concat = mx.symbol.Concat(*[roi_pool3_norm_scale, roi_pool4_norm_scale, roi_pool5_norm_scale], name="roi_pool_concat")`

And initilization in train_end2end.py:
`arg_params['roi_pool3_scale'] = mx.nd.full(arg_shape_dict['roi_pool3_scale'], 2)
 arg_params['roi_pool4_scale'] = mx.nd.full(arg_shape_dict['roi_pool4_scale'], 2)
 arg_params['roi_pool5_scale'] = mx.nd.full(arg_shape_dict['roi_pool5_scale'], 2)`

How do you these code running in Mxnet can't work well as in caffe? I very appreciate you can reply.

**Comment by zhreshold** [comment]
I don't get your point. How is it don't work well as in caffe?

**Comment by coolbrain** [comment]
@zhreshold  Inside-Outside Network is doing l2normalization and scale for each of conv3，conv4 and conv5 in VGG16  and concating them. I detect the text of ICDAR2013 dataset by this network, the F-measure of Caffe is 88%, but  82% in Mxnet. I found that the mean of scale value in caffe changes from 2 to 1.68, but hardly changes in Mxnet in training phrase. So I set wd_decay = 0.1  of scale in mxnet, the mean of scale changes a little.  in the meantime, the F-meansure of mxnet improves a little.

**Comment by zhreshold** [comment]
I once discovered the same issue with scaling, adding wd_mult to attr will fix it, though I don't know if it exactly how caffe version getting the smaller scale. 
The scaling is a trivial issue, I'd recommand you to check other stuffs, 6% is a lot and I don't think it is simply the different scaling problem.
Anyway, feel free to discuss what you have found resolving those problems!

**Comment by coolbrain** [comment]
@zhreshold  I just set the scale name end by "_weight", the scale value changes right by epochs. So I thought the backpropagation of scale has no problem.But the performance of mxnet remains 82% compared to 88% in caffe. Could you make sure that the backpropagation 
 of Normalize layer is same as L2Normalize in Mxnet?

**Comment by zhreshold** [comment]
It is confirmed by converting the model from caffe

**Comment by coolbrain** [comment]
I read the L2normalization code you wrote. This is the back propagation code which was selected channel. I found that mxnet code is a little different from the Caffe implementation.   
 `Tensor<xpu, 3> data = out_data[l2_normalization::kOut]
        .get_with_shape<xpu, 3, real_t>(dshape, s);`
x is input data; y is output data.
`Assign(grad_in, req[l2_normalization::kData],
        (grad_out - data * broadcast_with_axis(temp, 0, orig_shape[1])) /
        broadcast_with_axis(norm, 0, `orig_shape[1]));`
I think the code should be modified as this:
`Assign(grad_in, req[l2_normalization::kData],
        (grad_out - data * broadcast_with_axis(temp, 0, orig_shape[1])**/ broadcast_with_axis(norm, 0, orig_shape[1])**) /broadcast_with_axis(norm, 0, orig_shape[1]));`

or 
`Tensor<xpu, 3> data = in_data[l2_normalization::kData]
        .get_with_shape<xpu, 3, real_t>(dshape, s);
Assign(grad_in, req[l2_normalization::kData],
              (grad_out - (data * broadcast_with_axis(temp, 0, orig_shape[1]))
                 / broadcast_with_axis(F<mxnet::op::mshadow_op::square>(norm), 0, orig_shape[1])) /
               broadcast_with_axis(norm, 0, orig_shape[1]));`

**Comment by zhreshold** [comment]
The correct form is to divide the norm only once in l2_normalize,
we can reproduce it by compose a series of operators like this
```
sum = mx.sym.sum(mx.sym.square(data), axis=axis, keepdims=True)
inv_norm = mx.sym.rsqrt(sum + 1e-12)
out = mx.sym.broadcast_mul(data, inv_norm)
```
The gradients are properly handled, and you can compare the results.

**Comment by zhreshold** [comment]
<img width="300" alt="screen shot 2017-08-24 at 9 49 13 pm" src="https://user-images.githubusercontent.com/3307514/29700007-33fe9e38-8916-11e7-9f90-3cb3db78f3cc.png">
norm is squre root

**Comment by coolbrain** [comment]
I derived the propagation again,  It's like this.
![capture](https://user-images.githubusercontent.com/3873891/29717093-82ce53a8-89e0-11e7-8cab-0b7e7d04454e.PNG)
How do you think of two version of this implementation?
Your implementation:

`Tensor<xpu, 3> data = out_data[l2_normalization::kOut]
        .get_with_shape<xpu, 3, real_t>(dshape, s);
temp = reduce_with_axis<red::sum, false>(grad_out * data, 1);
      Assign(grad_in, req[l2_normalization::kData],
        (grad_out - data * broadcast_with_axis(temp, 0, orig_shape[1])) /
        broadcast_with_axis(norm, 0, orig_shape[1]));`

My implementation:
`Tensor<xpu, 3> data = in_data[l2_normalization::kData]
        .get_with_shape<xpu, 3, real_t>(dshape, s);
Assign(grad_in, req[l2_normalization::kData], (grad_out - (data * broadcast_with_axis(temp, 0, orig_shape[1])) / broadcast_with_axis(F<mxnet::op::mshadow_op::square>(norm), 0, orig_shape[1])) / broadcast_with_axis(norm, 0, orig_shape[1]));`

**Comment by zhreshold** [comment]
Mathematically they are the same, implementation-wise, I'd use the current one since it's shorter and faster.

**Comment by coolbrain** [comment]
I test the network in two implementation separately. But the performance of my implementation is lower than your implementation. In the meantime, the performance of your implementation is lower than the Caffe implementation. It doesn't conform to the derivation of mathematics. I am confused in this problem.

**Comment by zhreshold** [comment]
Did you get better performance using your implementation? I am curious as well.

**Comment by coolbrain** [comment]
It is the same. I can confirm the implementation of L2Normalization has no problem. I have figure out the problem in other places. Thanks for your help!

**Comment by zhreshold** [comment]
@coolbrain Do you mind telling me what problem did you figured out? I recently found the training part broken again a little bit after refactoring some part.

**Comment by coolbrain** [comment]
I haven't used mxnet-ssd you wrote. I used mx-rcnn to do detection.
